server/src/main/java/org/opensearch/cluster/routing/remote/RemoteRoutingTableService.java

```diff
@@ -42,6 +77,142 @@ public RemoteRoutingTableService(Supplier<RepositoriesService> repositoriesServi
+    ) {
+        final Map<String, ClusterMetadataManifest.UploadedIndexMetadata> allUploadedIndicesRouting = previousManifest.getIndicesRouting()
+            .stream()
+            .collect(Collectors.toMap(ClusterMetadataManifest.UploadedIndexMetadata::getIndexName, Function.identity()));
+
+        indicesRoutingUploaded.forEach(
+            uploadedIndexRouting -> allUploadedIndicesRouting.put(uploadedIndexRouting.getIndexName(), uploadedIndexRouting)
+        );
+        indicesRoutingToDelete.forEach(allUploadedIndicesRouting::remove);
+
+        return new ArrayList<>(allUploadedIndicesRouting.values());
+    }
+
+    private void uploadIndex(
+        IndexRoutingTable indexRouting,
+        String fileName,
+        BlobContainer blobContainer,
+        ActionListener<Void> completionListener
+    ) throws IOException {
+        RemoteIndexRoutingTable indexRoutingInput = new RemoteIndexRoutingTable(indexRouting);
+        BytesReference bytesInput = null;
+        try (BytesStreamOutput streamOutput = new BytesStreamOutput()) {
+            indexRoutingInput.writeTo(streamOutput);
+            bytesInput = streamOutput.bytes();
+        } catch (IOException e) {
+            throw new IOException("Failed to serialize IndexRoutingTable. ", e);
+        }
+
+        if (blobContainer instanceof AsyncMultiStreamBlobContainer == false) {
+            try {
+                blobContainer.writeBlob(fileName, bytesInput.streamInput(), bytesInput.length(), true);
+                completionListener.onResponse(null);
+            } catch (IOException e) {
+                logger.error("Failed to write IndexRoutingTable to remote store. ", e);
+                completionListener.onFailure(e);
+            }
+            return;
+        }
+
+        try (IndexInput input = new ByteArrayIndexInput("indexrouting", BytesReference.toBytes(bytesInput))) {
+            try (
+                RemoteTransferContainer remoteTransferContainer = new RemoteTransferContainer(
+                    fileName,
+                    fileName,
+                    input.length(),
+                    true,
+                    WritePriority.URGENT,
+                    (size, position) -> new OffsetRangeIndexInputStream(input, size, position),
+                    null,
+                    false
+                )
+            ) {
+                ((AsyncMultiStreamBlobContainer) blobContainer).asyncBlobUpload(
+                    remoteTransferContainer.createWriteContext(),
+                    completionListener
+                );
+            } catch (IOException e) {
+                logger.error("Failed to write IndexRoutingTable to remote store. ", e);
+                completionListener.onFailure(e);
+            }
+        } catch (IOException e) {
+            logger.error("Failed to create transfer object for IndexRoutingTable for remote store upload. ", e);
+            completionListener.onFailure(e);
+        }
```

Comment:
Can we use these interfaces https://github.com/opensearch-project/OpenSearch/pull/13785/files for remote transfers?

Reply:
Will take this refactoring as a separate change, creates issue to track this: https://github.com/opensearch-project/OpenSearch/issues/14067